simplicity-sys/src/ffi.rs

```diff
@@ -0,0 +1,480 @@
+        /// that specification
+        pub fn mkJetCMR(imr: *const u32) -> CSha256Midstate;
+
+        /// Compute the CMR of a jet of scribe(v) : ONE |- TWO^(2^n) that outputs a given
+        /// bitstring
+        pub fn computeWordCMR(value: *const CBitstring, n: size_t) -> CSha256Midstate;
+
+        /// Given a well-formed dag[i + 1], set the `cmr` field of every node in `dag`
+        pub fn computeCommitmentMerkleRoot(dag: *mut CDagNode, i: size_t);
+
+        /// Given a well-typed dag representing a Simplicity expression, compute
+        /// the annotated Merkle roots of all subexpressions.
+        pub fn computeAnnotatedMerkleRoot(
+            analyses: *mut CAnalyses,
+            dag: *const CDagNode,
+            ty: *const CType,
+            len: size_t,
+        );
+
+        /// Verifies that the 'dag' is in canonical order, meaning that nodes
+        /// under the left branches have lower indices than nodes under
+        pub fn verifyCanonicalOrder(dag: *const CDagNode, len: size_t) -> SimplicityErr;
+
+        /// Fills in the 'WITNESS' nodes of a 'dag' with the data from 'witness'
+        pub fn fillWitnessData(
+            dag: *mut CDagNode,
+            type_dag: *mut CType,
+            len: size_t,
+            witness: CBitstring,
+        ) -> SimplicityErr;
+
+        /// Computes the identity Merkle roots of every subexpression in a well-typed 'dag' with witnesses    .
+        pub fn verifyNoDuplicateIdentityRoots(
+            imr: *mut CSha256Midstate,
+            dag: *const CDagNode,
+            type_dag: *const CType,
+            len: size_t,
+        ) -> SimplicityErr;
+    }
+}
+
+pub mod deserialize {
+    use super::bitstream::CBitstream;
+    use super::bitstring::CBitstring;
+    use super::dag::{CCombinatorCounters, CDagNode};
+    use super::SimplicityErr;
+
+    extern "C" {
+        pub fn decodeMallocDag(
+            dag: *mut *mut CDagNode,
+            combinator_counters: *mut CCombinatorCounters,
+            stream: *mut CBitstream,
+        ) -> i32;
+
+        pub fn decodeWitnessData(
+            witness: *mut CBitstring,
+            stream: *mut CBitstream,
+        ) -> SimplicityErr;
+    }
+}
+
+pub mod eval {
+    use super::dag::CDagNode;
+    use super::ty::CType;
+    use super::SimplicityErr;
+    use super::{ubounded, UWORD};
+    use libc::{c_uchar, c_void, size_t};
+    use std::ptr;
+
+    pub const CHECK_NONE: c_uchar = 0;
+    pub const CHECK_EXEC: c_uchar = 0x10;
+    pub const CHECK_CASE: c_uchar = 0x60;
+    pub const CHECK_ALL: c_uchar = 0xFF;
+
+    extern "C" {
+        /// Run the Bit Machine on the well-typed Simplicity expression 'dag[len]'.
+        pub fn evalTCOExpression(
+            anti_dos_checks: c_uchar,
+            output: *mut UWORD,
+            outputSize: ubounded,
+            input: *const UWORD,
+            inputSize: ubounded,
+            dag: *const CDagNode,
+            type_dag: *const CType,
```

Comment:
`type_dag` should be *const. For some reason which I did not investigate, it is called into writeValue function that is indirectly called by this.